aeppl/logpdf.py

```diff
@@ -0,0 +1,455 @@
+from functools import singledispatch
+
+import aesara.tensor as at
+import aesara.tensor.random.basic as arb
+import numpy as np
+from aesara.assert_op import Assert
+from aesara.graph.op import Op
+from aesara.tensor.slinalg import Cholesky, solve_lower_triangular
+from aesara.tensor.var import TensorVariable
+
+# from aesara.tensor.xlogx import xlogy0
+
+cholesky = Cholesky(lower=True, on_error="nan")
+
+
+def betaln(x, y):
+    return at.gammaln(x) + at.gammaln(y) - at.gammaln(x + y)
+
+
+def binomln(n, k):
+    return at.gammaln(n + 1) - at.gammaln(k + 1) - at.gammaln(n - k + 1)
+
+
+def xlogy0(m, x):
+    # TODO: This should probably be a basic Aesara stabilization
+    return at.switch(at.eq(x, 0), at.switch(at.eq(m, 0), 0.0, -np.inf), m * at.log(x))
+
+
+def logpdf(rv_var, obs, **kwargs):
+    """Create a graph for the log-density/mass of a ``RandomVariable``."""
+    return _logpdf(rv_var.owner.op, obs, *rv_var.owner.inputs, **kwargs)
+
+
+@singledispatch
+def _logpdf(
+    op: Op,
+    value: TensorVariable,
+    *inputs: TensorVariable,
+    **kwargs,
+):
+    """Create a graph for the log-density/mass of a ``RandomVariable``.
+
+    This function dispatches on the type of ``op``, which should be a subclass
+    of ``RandomVariable``.  If you want to implement new density/mass graphs
+    for a ``RandomVariable``, register a new function on this dispatcher.
+
+    The default returns a graph producing only zeros.
+
+    """
+    raise NotImplementedError()
+
+
+@_logpdf.register(arb.UniformRV)
+def uniform_logpdf(op, value, *inputs, **kwargs):
+    lower, upper = inputs[3:]
+    return at.switch(
+        at.bitwise_and(at.ge(value, lower), at.le(value, upper)),
+        at.fill(value, -at.log(upper - lower)),
+        -np.inf,
+    )
+
+
+@_logpdf.register(arb.NormalRV)
+def normal_logpdf(op, value, *inputs, **kwargs):
+    mu, sigma = inputs[3:]
+    res = (
+        -0.5 * at.pow((value - mu) / sigma, 2)
+        - at.log(np.sqrt(2.0 * np.pi))
+        - at.log(sigma)
+    )
+    res = Assert("sigma > 0")(res, at.all(at.gt(sigma, 0.0)))
```

Comment:
The Assets sound neat compared to the pymc bound. It also means we might be able to disable them if we know it's safe to on the spot during optimizations, no?

Reply:
Yes, they can be optimized out.